packages/wdio-browserstack-service/src/Percy/Percy-Handler.ts

```diff
@@ -0,0 +1,176 @@
+            PercyLogger.error(`Error while trying to auto capture Percy screenshot ${err}`)
+        }
+    }
+
+    async before () {
+        this._browser.percyCaptureMap = new PercyCaptureMap()
+    }
+
+    deferCapture(sessionName: string, eventName: string | null) {
+        /* Service doesn't wait for handling of browser commands so the below counter is used in teardown method to delay service exit */
+        this._percyScreenshotCounter += 1
+        this.percyDeferredScreenshots.push({ sessionName, eventName })
+    }
+
+    isDOMChangingCommand(args: BeforeCommandArgs): boolean {
+        if ((args.method as string) === 'POST') {
+            if (PERCY_DOM_CHANGING_COMMANDS_ENDPOINTS.includes((args.endpoint as string))) {
+                return true
+            } else if ((args.endpoint as string).includes('/session/:sessionId/element') && (args.endpoint as string).includes('click')) {
+                /* click element */
+                return true
+            } else if ((args.endpoint as string).includes('/session/:sessionId/element') && (args.endpoint as string).includes('clear')) {
+                /* clear element */
+                return true
+            } else if ((args.endpoint as string).includes('/session/:sessionId/execute') && args.body?.script) {
+                /* execute script sync / async */
+                return true
+            } else if ((args.endpoint as string).includes('/session/:sessionId/touch')) {
+                /* Touch action for Appium */
+                return true
+            }
+        } else if ((args.method as string) === 'DELETE' && (args.endpoint as string) === '/session/:sessionId') {
+            return true
+        }
+        return false
+    }
+
+    async cleanupDeferredScreenshots() {
+        this.isPercyCleanupProcessingUnderway = true
+        for await (const entry of this.percyDeferredScreenshots) {
+            await this.percyAutoCapture(entry.eventName, entry.sessionName)
+        }
+        this.percyDeferredScreenshots = []
+        this.isPercyCleanupProcessingUnderway = false
+    }
+
+    async #sleep(ms: number) {
+        return new Promise(resolve => setTimeout(resolve, ms))
+    }
+
+    async browserBeforeCommand (args: BeforeCommandArgs) {
+        try {
+            if (this.isDOMChangingCommand(args)) {
+                do {
+                    await this.#sleep(1000)
+                } while (this.percyScreenshotInterval)
+                this.percyScreenshotInterval = setInterval(async () => {
+                    if (!this.isPercyCleanupProcessingUnderway) {
+                        clearInterval(this.percyScreenshotInterval)
+                        await this.cleanupDeferredScreenshots()
+                        this.percyScreenshotInterval = null
+                    }
+                }, 1000)
+            }
+        } catch (err: any) {
+            PercyLogger.error(`Error while trying to cleanup deferred screenshots ${err}`)
+        }
```

Comment:
Return early if possible.
```suggestion
        if (this.isDOMChangingCommand(args)) {
            return
        }

        try {
            do {
                await this.#sleep(1000)
            } while (this.percyScreenshotInterval)
            this.percyScreenshotInterval = setInterval(async () => {
                if (this.isPercyCleanupProcessingUnderway) {
                    return
                }
                clearInterval(this.percyScreenshotInterval)
                await this.cleanupDeferredScreenshots()
                this.percyScreenshotInterval = null
            }, 1000)
        } catch (err: any) {
            PercyLogger.error(`Error while trying to cleanup deferred screenshots ${err}`)
        }
```

What is the purpose of the do/while loop here?

Reply:
Since browser command events don't wait on the completion of listeners, multiple DOM changing commands might end up registering a `setInterval` to clean up deferred screenshots but we only want the first command to do so hence the do while loop makes successive browser commands wait until the previously registered `setInterval` is completed before registering a `setInterval` of their own